### PR TITLE
[V1][Spec Decode] Share input embedding of target model with EAGLE draft model to free ~1GB for llama 3 model

### PR DESCRIPTION
**Pull request by ekagra-ranjan**
This PR:
* Fixes: https://github.com/vllm-project/vllm/pull/16035#discussion_r2033572712
* Frees ~1GB GPU memory for llama 3 model
* Add output logging for sanity check in eagle offline bench
* Tested by using my hacky metric [PR](https://github.com/ekagra-ranjan/vllm/pull/1/commits) and ensuring AL is high.

### CODE REVIEW

File: vllm/v1/spec_decode/eagle.py [hunk 1]
```diff
@@ -235,17 +235,18 @@ def load_model(self, target_model: nn.Module) -> None:
                     model_config=draft_model_config,
                     start_layer_id=target_layer_num).to(target_device)
 
-        loaded_weights = self.model.load_weights(
+        _ = self.model.load_weights(
```
Comment (WoosukKwon) [inline vllm/v1/spec_decode/eagle.py:238]:
```suggestion
        self.model.load_weights(
```

**Review by WoosukKwon — COMMENTED** [review]
@ekagra-ranjan Thanks for the PR!

One issue with the PR is that it assumes PP=1. Can you please handle PP > 1 as well (at least for llama)?

**Comment by WoosukKwon** [comment]
@ekagra-ranjan Could you please update the PR? If handling PP is tricky, you can simply check the `pipeline_parallel_size` and raise an error if it's not 1 (for now). We can fix it later.

File: vllm/model_executor/models/llama_eagle.py [hunk 2]
```diff
@@ -142,14 +153,12 @@ def forward(
     def load_weights(self, weights: Iterable[Tuple[str, torch.Tensor]]):
         loader = AutoWeightsLoader(
             self,
-            skip_prefixes=(["lm_head."]
-                           if self.config.tie_word_embeddings else None),
```
Comment (ekagra-ranjan) [inline vllm/model_executor/models/llama_eagle.py:146]: eagle model def doesnt have lm_head nor the weights to removed it

**Comment by ekagra-ranjan** [comment]
@WoosukKwon Done! For PP > 1, the target embed would be on rank 0 whereas the drafter will run on last rank so the drafter's embed cannot be shared with target. The current code will expect the embed weights to be present in draft checkpoint when using PP and raise an exception if that's not the case.

File: vllm/model_executor/models/llama_eagle.py (hunk 2 above)
Comment (WoosukKwon) [reply vllm/model_executor/models/llama_eagle.py:146]: @ekagra-ranjan Do you mean EAGLE1 doesn't have the LM head? I'm wondering because some EAGLE3 weights do include the LM head.

File: vllm/v1/spec_decode/eagle.py [hunk 3]
```diff
@@ -302,12 +303,25 @@ def load_model(self, target_model: nn.Module) -> None:
...
-                logger.info(
-                    "Loading EAGLE embedding weights from the target model.")
-                self.model.model.embed_tokens = target_model.model.embed_tokens
+
+        # share embed_tokens with the target model if needed
+        if get_pp_group().world_size == 1:
+            assert "model.embed_tokens.weight" not in loaded_weights, \
+            "For PP = 1, Eagle draft should share embed with target model"
+            logger.info(
+                "Loading EAGLE embedding weights from the target model.")
```
Comment (WoosukKwon) [inline vllm/v1/spec_decode/eagle.py:312]:
```suggestion
            logger.info(
                "The EAGLE head shares the same vocab embedding with the target model.")
```

File: vllm/v1/spec_decode/eagle.py [hunk 4]
```diff
@@ -302,12 +303,25 @@ def load_model(self, target_model: nn.Module) -> None:
...
+            self.model.model.embed_tokens = target_model.model.embed_tokens
         else:
+            assert "model.embed_tokens.weight" in loaded_weights, \
+            "For PP > 0, Eagle draft checkpoint should its own copy of "
```
Comment (WoosukKwon) [inline vllm/v1/spec_decode/eagle.py:316]:
```suggestion
            "For PP > 1, Eagle draft checkpoint should its own copy of "
```

File: vllm/v1/spec_decode/eagle.py [hunk 5]
```diff
@@ -302,12 +303,25 @@ def load_model(self, target_model: nn.Module) -> None:
...
+            " the model.embed_tokens.weight"
+            logger.info("EAGLE embedding weights are already loaded.")
```
Comment (WoosukKwon) [inline vllm/v1/spec_decode/eagle.py:318]:
```suggestion
            logger.info("Since PP > 1, the EAGLE head loaded its own vocab embedding weights instead of sharing them with the target model.")
```

**Review by WoosukKwon — COMMENTED** [review]
@ekagra-ranjan Left some minor comments. Please check them out.

File: vllm/model_executor/models/llama_eagle.py (hunk 2 above)
Comment (ekagra-ranjan) [reply vllm/model_executor/models/llama_eagle.py:146]: EAGLE1 reuses the lm_head of target model for each spec step whereas EAGLE3 does not. For e.g., 

`yuhuili/EAGLE-LLaMA3-Instruct-8B` has these weights

```
Number of weights: 10
Key: layers.0.self_attn.q_proj.weight, Shape: torch.Size([4096, 4096]), Dtype: torch.float16
Key: layers.0.self_attn.k_proj.weight, Shape: torch.Size([1024, 4096]), Dtype: torch.float16
Key: layers.0.self_attn.v_proj.weight, Shape: torch.Size([1024, 4096]), Dtype: torch.float16
Key: layers.0.self_attn.o_proj.weight, Shape: torch.Size([4096, 4096]), Dtype: torch.float16
Key: layers.0.mlp.gate_proj.weight, Shape: torch.Size([14336, 4096]), Dtype: torch.float16
Key: layers.0.mlp.up_proj.weight, Shape: torch.Size([14336, 4096]), Dtype: torch.float16
Key: layers.0.mlp.down_proj.weight, Shape: torch.Size([4096, 14336]), Dtype: torch.float16
Key: layers.0.post_attention_layernorm.weight, Shape: torch.Size([4096]), Dtype: torch.float16
Key: embed_tokens.weight, Shape: torch.Size([128256, 4096]), Dtype: torch.float16
Key: fc.weight, Shape: torch.Size([4096, 8192]), Dtype: torch.float16
```

EAGLE 3's lm_head is not the same as the target model. It's noted in this PR as well https://github.com/vllm-project/vllm/pull/16937#issue-3008960986

**Comment by WoosukKwon** [comment]
@ekagra-ranjan Please fix the lint errors.

**Comment by singh-git10** [comment]
@ekagra-ranjan @WoosukKwon I believe the scenario where the EAGLE-3 draft model has different embedding weights than the target model is not being properly handled in the current implementation. This issue specifically applies to the EAGLE-3 head for the Llama 3.3 70B model.([yuhuili/EAGLE3-LLaMA3.3-Instruct-70B](https://huggingface.co/yuhuili/EAGLE3-LLaMA3.3-Instruct-70B/tree/main).

**Comment by ekagra-ranjan** [comment]
@singh-git10 - its being addressed here: https://github.com/vllm-project/vllm/pull/19033
